### PR TITLE
fix: Add retry loop for secret client if initial token is invalid

### PR DESCRIPTION
**Pull request by bnevis-i**
Address a race condition where the container may be started before
security-secrets-setup has created the secrets token for the service.

Signed-off-by: Bryon Nevis <bryon.nevis@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
When user performs

```sh
docker-compose -f docker-compose-blah.yml up -d
````

A number of core services such as support-logging, core-metadata, core-data, and others will fail to start with a message:

```
level=ERROR ts=2020-03-20T16:33:16.413174639Z app=edgex-core-command source=secret.go:67 msg="unable to create SecretClient: HTTP response with status code 403, message: failed to lookup token"
```

Issue Number:
- https://github.com/edgexfoundry/developer-scripts/issues/230
- https://github.com/edgexfoundry/edgex-go/issues/2404

## What is the new behavior?

Expect to see messages similar to the following which shows that retries are happening.  You will either get secrets-token.json doesn't exist, OR will fail to lookup token, depending on whether or not /tmp/edgex/secrets is empty.

```
level=INFO ts=2020-03-20T19:02:58.651792922Z app=edgex-core-data source=secret.go:60 msg="Creating SecretClient"
level=INFO ts=2020-03-20T19:02:58.65186062Z app=edgex-core-data source=secret.go:71 msg="Attempting to create secret client (attmpt 1 of 11)"
level=ERROR ts=2020-03-20T19:31:52.47329444Z app=edgex-core-data source=secret.go:75 msg="unable to parse secret store configuration: open /tmp/edgex/secrets/edgex-core-data/secrets-token.json: no such file or directory"
level=ERROR ts=2020-03-20T19:02:58.774292175Z app=edgex-core-data source=secret.go:86 msg="Retryable failure (after 1s delay) while creating SecretClient: HTTP response with status code 503, message: failed to lookup token"
level=INFO ts=2020-03-20T19:02:59.774681651Z app=edgex-core-data source=secret.go:71 msg="Attempting to create secret client (attmpt 2 of 11)"
level=ERROR ts=2020-03-20T19:31:52.47329444Z app=edgex-core-data source=secret.go:75 msg="unable to parse secret store configuration: open /tmp/edgex/secrets/edgex-core-data/secrets-token.json: no such file or directory"
level=ERROR ts=2020-03-20T19:02:59.809673211Z app=edgex-core-data source=secret.go:86 msg="Retryable failure (after 1s delay) while creating SecretClient: HTTP response with status code 403, message: failed to lookup token"
level=INFO ts=2020-03-20T19:03:00.810353941Z app=edgex-core-data source=secret.go:71 msg="Attempting to create secret client (attmpt 3 of 11)"
level=ERROR ts=2020-03-20T19:31:52.47329444Z app=edgex-core-data source=secret.go:75 msg="unable to parse secret store configuration: open /tmp/edgex/secrets/edgex-core-data/secrets-token.json: no such file or directory"
level=ERROR ts=2020-03-20T19:03:00.846627763Z app=edgex-core-data source=secret.go:86 msg="Retryable failure (after 1s delay) while creating SecretClient: HTTP response with status code 403, message: failed to lookup token"
level=INFO ts=2020-03-20T19:03:01.846973046Z app=edgex-core-data source=secret.go:71 msg="Attempting to create secret client (attmpt 4 of 11)"
level=INFO ts=2020-03-20T19:03:01.881978472Z app=edgex-core-data source=secret.go:82 msg="Created SecretClient"

```

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Are there any specific instructions or things that should be known prior to reviewing?

In order to test, modify go.mod in edgex-go

```
module github.com/edgexfoundry/edgex-go

replace github.com/edgexfoundry/go-mod-bootstrap => github.com/bnevis-i/go-mod-bootstrap v0.0.0-20200320192440-3829504ae15b

require (
```

and rebuild docker containers.  Update developer scripts to refer to newly built containers.  Bring up compose stack as above.

## Other information

### CODE REVIEW

**Comment by tingyuz** [comment]
With the recent change of #58 the proxy-setup has same issue as described here. This will be a pre-required PR for proxy-setup to move forward.

**Comment by bnevis-i** [comment]
@AnthonyMBonafide are you open to help review this?

**Comment by AnthonyMBonafide** [comment]
Yes, l will take a look

File: bootstrap/handlers/secret/secret.go [hunk 1]
```diff
@@ -52,16 +53,38 @@ func (s *SecretProvider) BootstrapHandler(
...
+			} else {
+				secretClient, err := client.NewVault(ctx, secretConfig, lc).Get(configuration.GetBootstrap().SecretStore)
+				if err == nil {
+					s.secretClient = secretClient
+					lc.Info("Created SecretClient")
+					break
+				}
+			}
+
+			lc.Error(fmt.Sprintf("Retryable failure (after %s delay) while creating SecretClient: %s", secretStoreConfig.RetryWaitPeriod, err.Error()))
```
Comment (AnthonyMBonafide) [inline bootstrap/handlers/secret/secret.go]: `err` at this point may be `nil`. Should we add a `nil` check or move this into the previous `if-else` statement?

File: bootstrap/handlers/secret/secret.go [hunk 2]
```diff
@@ -52,16 +53,38 @@ func (s *SecretProvider) BootstrapHandler(
...
+			time.Sleep(retryWaitPeriod)
+		}
 
 		if err != nil {
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: err is declared on line 58 and is updated at line 73 and 77.  If the retry loop exits and err is still set we failed to initialize the secretclient and give up.  If the vault client is successfully initialized, err is set to nil, we break out of the loop, and skip over the bailout and proceed with normal execution.

File: bootstrap/handlers/secret/secret.go [hunk 3]
```diff
@@ -52,16 +53,38 @@ func (s *SecretProvider) BootstrapHandler(
...
+
+		configuration := container.ConfigurationFrom(dic.Get)
+		secretStoreConfig := configuration.GetBootstrap().SecretStore
+		retryWaitPeriod, err := time.ParseDuration(secretStoreConfig.RetryWaitPeriod)
+		if err != nil {
+			lc.Error(fmt.Sprintf("unable to parse RetryWaitPeriod: %s", err.Error()))
+			return false
+		}
+
+		for i := 1; i <= secretStoreConfig.AdditionalRetryAttempts+1; i++ {
```
[the remark this answers is not in the document]
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: Probably.  I will try to find another bootstrap handler that uses it and see what the other code is doing.

File: bootstrap/handlers/secret/secret.go (hunk 1 above)
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: if err is nil at line 74, we execute the else branch.
if err is nil at line 78, we break out of the loop and never get to line 85

What is the scenario where error is nil when we get here?

File: bootstrap/handlers/secret/secret.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: Disregard. Whole new implementation.

File: bootstrap/handlers/secret/secret.go (hunk 1 above)
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: Disregard. Whole new implementation.

File: bootstrap/handlers/secret/secret.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bnevis-i) [reply bootstrap/handlers/secret/secret.go]: Disregard. Whole new implementation.

**Comment by bnevis-i** [comment]
Complete rewrite to use the startupTimer instead.  Please re-review.

File: bootstrap/handlers/secret/secret.go [hunk 4]
```diff
@@ -40,28 +40,50 @@ func NewSecret() *SecretProvider {
...
+				lc.Info("Attempting to create secretclient")
+				secretClient, err = client.NewVault(ctx, secretConfig, lc).Get(configuration.GetBootstrap().SecretStore)
+				if err == nil {
+					s.secretClient = secretClient
+					lc.Info("Created SecretClient")
+					break
+				}
+			}
+
+			if err != nil {
```
Comment (AnthonyMBonafide) [inline bootstrap/handlers/secret/secret.go]: This conditional looks like it will always be `true`, can this be removed and just add the logging statement?

**Pull request merged**